[FLOSS H3] simplify infos/status

janldeboer opened the pull request:
Reopening the pull request #3590 after rebasing to the current master.

fixes #666
fixes #3504

<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [x] add a line in `doc/news/_preparation_next_release.md`
  - [x] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you,
but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For docu fixes, spell checking, and similar none of these points below
need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

Commit:
simplify infos/status

fixes #666
fixes #3504

Janldeboer commented:
Thera are currently two tests that fail:
```sh
The following tests FAILED:
[877](https://github.com/ElektraInitiative/libelektra/actions/runs/3642273997/jobs/6149183772#step:10:878)
	 17 - testtool_specreader (Failed)
[878](https://github.com/ElektraInitiative/libelektra/actions/runs/3642273997/jobs/6149183772#step:10:879)
	181 - testscr_check_kdb_internal_check (Failed)
```
I'm trying to fix them, but I have difficulties understanding what they even check and how that relates to this changes.

kodebach commented:
`testtool_specreader` tests the code that reads the `infos/status` data. This includes

https://github.com/ElektraInitiative/libelektra/blob/fda27d8b7db5fb25a864290d917d06bcdbaf55dd/src/libs/tools/tests/testtool_specreader.cpp#L132

But AFAICT `tested` no longer exists as a status.

The `testscr_check_kdb_internal_check` runs `Plugin::check`

https://github.com/ElektraInitiative/libelektra/blob/fda27d8b7db5fb25a864290d917d06bcdbaf55dd/src/libs/tools/src/plugin.cpp#L157

on (almost) every plugin. That also includes some checks for `infos/status`. You changed the allowed `infos/status` values (`PluginDatabase::statusMap`), but didn't update the READMEs of the plugins.

Janldeboer commented:
Okay, that makes sense, thank you for your help!
As I didn't come up with the changes myself but just reopened an old PR, I'm unsure how exactly this was planned to be implemented.

I currently see two options to resolve the issues:
1. Change the `README`s of all plugins in question
2. Change the allowed values of `infos/status` so that all old values are still supported

Personally I think the 2. option is more appropriate, do you agree with that?

markus2330 reviewed (changes requested):
Small updates from changed terminology and decisions.

markus2330 commented:
> Change the READMEs of all plugins in question

Changing all READMEs is needed to fix #666, which purpose is to update/simplify what is allowed in `infos/status`. So obviously the outdated parts of `infos/status` must be get rid of.

Please fix `scripts/dev/update-infos-status`, then most of these updates should be quite easy.

Replacing "nodep" and adding the storage features is something bigger, please start with applying the changes of `infos/status` to the plugins. I updated #666 with the tasks. So for this PR, it is better to keep nodep (add it back to [src/libs/tools/include/plugincontract.h](https://github.com/ElektraInitiative/libelektra/pull/4758/files#diff-1afabe6b235d65073682d035ac9842c735f62be0a95911e689e90fce13c62309)).

janldeboer committed:
doc: Fix grammar mistake

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

janldeboer committed:
src: Use `hook` instead of `global`

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

janldeboer committed:
src: Use `internal` instead of `private`

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

janldeboer committed:
doc: Updated release notes

janldeboer committed:
scripts: Fix typo

janldeboer committed:
src: Restyled by clang-format

janldeboer committed:
src: Fixed spelling

janldeboer committed:
src: Fix spelling

janldeboer committed:
src: Fix spelling

Janldeboer commented:
Parsing the contract to retrieve all available tags is not working anymore, as the `CONTRACT.ini` isn't compiled before running the regex, so the tags are available at that point in time.
Shall I just parse the `plugincontract.h` directly or do we have to compile the CONTRACT.ini so that we can keep using it for the test?

Janldeboer commented:
Looking at the tags which aren't available anymore (except `nodep`), I wonder how to replace which ones and which to remove:
concept -> experimental
coverage -> ?
limited -> experimental (?)
readonly -> delete (?)
nodoc -> delete (?)
writeonly -> delete (?)
tested -> tested/unit or tested/shell (?)
conformant -> ?
difficult -> experimental (?)
unfinished -> preview
final -> productive
memleak -> preview (?)
reviewed -> documented (?)
unittest -> tested/unit
specific -> ?
shelltest -> tested/shell
global -> ?
libc -> ?

I adjusted the `scripts/dev/update-infos-status` to allow deleting and replacing multiple tags with one command, so once we're settled on how the tags shall be changed it can quickly be applied.

For this, I removed the restraint that `--del`, `--set` and `--add` are mutual exclusive and I introduced a `--replace` flag, to be used like this:
`--replace oldtag1 newtag1 oldtag2 newtag2 ...`

Janldeboer commented:
Another thing I don't fully understand is where the `pluginfeatures.h` are being used.
The tags are included in the `CONTRACT.ini` in `[infos/features/storage]`, but part of the contract seems unused.

Janldeboer commented:
> Looking at the tags which aren't available anymore (except `nodep`), I wonder how to replace which ones and which to remove

@markus2330 I would appreciate any thoughts on this, as I don’t want to make this decision on my own. Thank you very much!

markus2330 commented:
You are doing a great job here! Thanks for diving in!

> Parsing the contract to retrieve all available tags is not working anymore, as the CONTRACT.ini isn't compiled before running the regex, so the tags are available at that point in time.

This is a problem to be fixed in CMake.

> Shall I just parse the plugincontract.h directly or do we have to compile the CONTRACT.ini so that we can keep using it for the test?

For update-infos-status it actually makes sense to parse what we have in the repo. (But if you already implemented otherwise, it is fine, too)

> Another thing I don't fully understand is where the pluginfeatures.h are being used.
> The tags are included in the CONTRACT.ini in [infos/features/storage], but part of the contract seems unused.

The plan is to use it in src/plugins/*/README.md of the storage plugins. But it might be too much work for you to find out for all the storage plugins which features they have.

About the replacements (unchanged lines are okay as you suggested):

```diff
concept -> experimental
+ coverage -> tested/unit // can be cross-checked by looking at https://doc.libelektra.org/coverage/master/debian-bullseye-full/
+ limited -> pluginfeatures.h // this is replaced by pluginfeatures.h
+ readonly -> pluginfeatures.h
+ nodoc -> documented
+ writeonly -> pluginfeatures.h
+ tested -> tested/unit or tested/shell (?) // tested/shell can be determined by grepping for TEST_README in the plugin's CMakeLists.txt
- conformant -> ? // no replacement
+ difficult -> experimental
unfinished -> preview
final -> productive
+ memleak -> memleak // memleak is unchanged, still exists
- reviewed -> documented // reviewed can be discarded
unittest -> tested/unit
+ specific -> pluginfeatures.h limited
shelltest -> tested/shell
+ global -> hook
- libc -> ? // discard
```

Janldeboer commented:
How should I handle the tags that are now covered by `pluginfeatures.h`?
As far as I see, the script doesn't allow be to bulk edit them.
Instead of the tags `readonly` and `writeonly` we now have the features `read`and `write`, so I suppose every plugin which is tagged with neither `readonly` or `writeonly` should have both features `read` and `write`?

Would it be a good approach to remove them first and add them back in a new PR, as we have to check the features used by each plugin anyway?

Checking it the features of all plugins is indeed too much work for me, at least until the deadline tomorrow.

janldeboer committed:
src, scripts: Migrating status tags to features

For this task I created a new script `scripts/dev/migrate-status-to-feature` that applied the required changes.

Note that the tags that are migrated are hardcoded, for reusing the script it might have to be changed.

janldeboer committed:
scripts: Read status values directly from file

Read the values allow for infos/status directly from `plugincontract.h` instead of `CONTRACT.ini`

janldeboer committed:
scripts: Make flags not mutually exclusive

janldeboer committed:
scripts: Don't check tags that will be deleted

As they will be removed anyways, the tags don't have to be valid

janldeboer committed:
scripts: Add `nodep` back to status values

janldeboer committed:
scripts: Improve check of unavailable tags

janldeboer committed:
scripts: Allow multiple flags to be processed

janldeboer committed:
scripts: Added replace flag to script

Janldeboer commented:
I successfully migrated all the tags that are features using my script.
However, I wasn't yet able to apply the remaining deletions / replacements:
First, in the list of removed tags I forgot `old`, `orphan` and `preview`.
I decided to remove `old` and `orphan` and replace `preview` by `concept`

For some reason, the script is completely deleting the READMEs from some of the plugins and I couldn't figure out why. I will join tomorrows consultation hour, so that hopefully we can manage to solve that issue.

Janldeboer commented:
```
old -> obsolete
orphan -> obsolete
preview -> concept
```

Janldeboer commented:
Update on how I will change the tags:
Deleting:
```
conformant
reviewed
libc
tested
nodoc
```
Replacing:
```
coverage -> tested/unit
difficult -> experimental
unfinished -> experimental
final -> productive
unittest -> tested/unit
shelltest -> tested/shell
global -> hook
preview -> experimental
orphan -> obsolete
old -> obsolete
```

janldeboer committed:
scripts: Check new tags when replacing

janldeboer committed:
src: Replace deprecated tags in plugin

Done by running the following script:
```
scripts/dev/update-infos-status --del conformant reviewed libc tested nodoc --replace coverage tested/unit difficult experimental unfinished experimental final productive unittest tested/unit shelltest tested/shell global hook preview experimental orphan obsolete old obsolete
```
which applies the following changes:
deleting:
```
conformant
reviewed
libc
tested
nodoc
```
replacing:
```
coverage -> tested/unit
difficult -> experimental
unfinished -> experimental
final -> productive
unittest -> tested/unit
shelltest -> tested/shell
global -> hook
preview -> experimental
orphan -> obsolete
old -> obsolete
```

restyled-commits committed:
src: Restyled by clang-format

Janldeboer commented:
I'm currently running in to a new problem:
Now that the status may contain a `/`,  the `generate_readme` function in `scripts/cmake/modules/LibAddMactos.cmake` doesn't work correctly anymore, as the `/` isn't parsed by the regex.
My attempts to fix the regex failed so far.

janldeboer committed:
scripts: Adjust `generate_readme` method

janldeboer committed:
tests: Replace deprecated tags

janldeboer committed:
scripts: patched formatting

janldeboer committed:
tests: Replaced deprecated tags

janldeboer committed:
tests: added rule to not check migration script

The python script 'scripts/dev/migrate-status-to-features' shall not be checked like a bash script

Janldeboer commented:
@markus2330 @flo91 
Unfortunately I wasn't able to finish the PR yesterday, as new problems keep coming up the further the CI gets.
I'm unhappy about the way I solved one particular issue:
The plugins `spec` and `c` had the tag `preview` before, which was replaced by `experimental`.
As all experimental plugins are excluded by default, the two were excluded as well, even though they are required for some tests.
I manually added them to the default plugin selection:
```
ALL;-EXPERIMENTAL;spec;c
```
Removing or changing the experimental tag (maybe to `concept`?) would probably be the cleaner solution.
What is your opinion on this?

Another more general problem that I noticed is that the check "FULL" is always failing, as Intel x86 builds on macOS are generally not supported anymore by the CI.

Janldeboer commented:
So far, two issue still occur that I have trouble to fix:
<details>
  <summary>libelektra-c.so not found</summary>
  ```
running test commands with parent key spec:/tests/script/gen/highlevel/commands
error: kdb gen failed: 
The command /home/jenkins/workspace/libelektra_PR-4758@3/build directory/bin/kdb gen terminated unsuccessfully with the info:
Was not able to load such a plugin!

Maybe you misspelled it, there is no such plugin or the loader has problems.
You might want to try to set LD_LIBRARY_PATH, use kdb-full or kdb-static.
Errors/Warnings during loading were:
 Sorry, 1 warning was issued ;(
 1: Module kdb issued the warning C01200:
	Installation: Dlopen failed. Could not load module libelektra-c.so. Reason: libelektra-c.so: cannot open shared object file: No such file or directory
	Mountpoint: /
	Configfile: 
	At: /home/jenkins/workspace/libelektra_PR-4758@3/src/libs/loader/dl.c:90

Please report the issue at https://issues.libelektra.org/
 ```
  This appeared before I added the `c` plugin to the default plugins manually and I hope that it won't appear anymore on the current run of the CI.
```
</details>

<details>
  <summary>selected default ... is not selected in PLUGINS</summary>
```
CMake Error at src/plugins/CMakeLists.txt:73 (message):
  selected default storage (dump) is not selected in PLUGINS, please change
  KDB_DEFAULT_STORAGE or PLUGINS


CMake Error at src/plugins/CMakeLists.txt:81 (message):
  selected default resolver (resolver_fm_hpu_b) is not selected in PLUGINS,
  please change KDB_DEFAULT_RESOLVER or PLUGINS
```
It seems to me like again the plugins are excluded due to changed tags, but in this case both plugins don't have the `experimental` tag. Maybe I'm able to fix this by digging deeper into why the plugins aren't available, but any help would be appreciated.
</details>

janldeboer committed:
src/tests: Replaced deprecated tags in tests

In some cases old tags remained in code snippets for testing

restyled-commits committed:
src: Restyled by clang-format

markus2330 commented:
> a lot of time to fix the issues because I always need to wait for the CI.

It shouldn't detain you from working on other PRs in between.

> As all experimental plugins are excluded by default, the two were excluded as well, even though they are required for some tests.

Then better keep the experimental status as it is. (The fix would be to properly exclude the tests if the plugins are not included in the build. This might be quite time consuming to find out how to do this).

Can you please add the information you have (error messages etc.) in #4820

> Removing or changing the experimental tag (maybe to concept?) would probably be the cleaner solution.

I agree. Simply let these three plugins without experimental.

> Another more general problem that I noticed is that the check "FULL" is always failing, as Intel x86 builds on macOS are generally not supported anymore by the CI.

Is this maybe #4793? If it is, please reopen the issue and write additional info there (error message etc.)

> selected default ... is not selected in PLUGINS

I don't understand this sentence.

markus2330 commented:
There is still a unit test failing. Please fix the CI problems so that we can get this merged.

Janldeboer commented:
@markus2330 I am aware of the remaining CI problem, but couldn't get behind why it came up.
It comes down to this function returning null:
`src/plugins/process/process.c`: `readKeySet` (lines 641 to 658, end of the file)
```
static KeySet * readKeySet (IoData * data, Key * errorKey)
{
	typedef int (*funserialize_t) (KeySet *, FILE *, Key *);

	funserialize_t funserialize = *(funserialize_t *) elektraInvokeGetFunction (data->dump, "funserialize");

	if (funserialize == NULL)
	{
		return NULL;
	}

	KeySet * ks = ksNew (0, KS_END);
	if (funserialize (ks, data->fromChild, errorKey) < 0)
	{
		return NULL;
	}
	return ks;
}
```
What surprises me, is that the test passed before. I think it's only failing since I replaced deprecated tags in the tests in commit [c10690d](https://github.com/ElektraInitiative/libelektra/pull/4758/commits/c10690d44c96e0567c9095c3f69a5d98c69f5cfc).

I will continue to investigate the root cause, any input is appreciated.

Janldeboer commented:
I also don't fully understand how to proceed with these instruction, as they seem to contradict each other:

> Then better keep the experimental status as it is.

> Simply let these three plugins be without experimental.

markus2330 commented on scripts/cmake/ElektraCache.cmake:
```diff
@@ -13,7 +13,7 @@ include (LibAddMacros)
 set (
 	PLUGINS
-	"ALL;-EXPERIMENTAL"
+	"ALL;-EXPERIMENTAL;spec;c;quickdump"
```
```suggestion
	"ALL;-EXPERIMENTAL
```

markus2330 commented on src/plugins/quickdump/README.md:
```diff
@@ -5,7 +5,7 @@
 - infos/recommends =
 - infos/placements = getstorage setstorage
-- infos/status = maintained compatible tested nodep libc preview
+- infos/status = maintained compatible nodep experimental
```
```suggestion
- infos/status = maintained compatible nodep
```

markus2330 commented on src/plugins/c/README.md:
```diff
@@ -5,7 +5,8 @@
 - infos/recommends =
 - infos/placements = getstorage setstorage
-- infos/status = maintained nodep libc writeonly preview nodoc
+- infos/status = maintained nodep experimental
```
```suggestion
- infos/status = maintained nodep
```

markus2330 commented on src/plugins/spec/README.md:
```diff
@@ -3,7 +3,7 @@
 - infos/needs =
 - infos/provides = check apply
-- infos/status = recommended productive nodep configurable global unfinished
+- infos/status = recommended productive nodep configurable hook experimental
```
```suggestion
- infos/status = recommended productive nodep configurable hook
```

markus2330 reviewed (changes requested):
> Then better keep the experimental status as it is.

Means that it should be like it was before the PR, see suggestions.

> src/plugins/process/process.c

@kodebach do you know which plugin is missing here causing this problem? (With minimal changes so that we can get this PR merged latest until Monday.)

kodebach commented:
The `process` plugin only depends on `dump`. That isn't properly checked in CMake, but it shouldn't be the issue, because `dump` is definitely included in the build. I'll try to find out what the issue is with loading `funserialize`.

I also agree that `spec`, `c` and `quickdump` should definitely not be experimental. All of them have existed for a while and are fully developed (apart from the upcoming changes to `spec`).

kodebach commented on src/plugins/process/testapp.sh:
```diff
@@ -74,7 +74,7 @@ system:/elektra/modules/process/infos/recommends
 
 $key string 44 42
```
This script prints `dump` format, so when you change the value of the key, you also need to update it's length accordingly:

```suggestion
$key string 44 48
```

@Janldeboer That should fix the issues with `testmod_process`

Janldeboer replied:
Thank you very much! I don't think I would have ever thought of that

janldeboer committed:
src: Fix key string length in `testapp.sh`

Co-authored-by: Klemens Böswirth <23529132+kodebach@users.noreply.github.com>

janldeboer committed:
src: Remove `experimental` from `quickdump`

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

janldeboer committed:
src: Remove `experimental` from `c` plugin

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

janldeboer committed:
src: Remove `experimental` from `spec` plugin

Co-authored-by: Markus Raab <markus2330@users.noreply.github.com>

markus2330 commented:
jenkins build libelektra please

markus2330 reviewed (approved):
LGTM, more changes better in a new PR

Janldeboer commented:
There is still one test failing:
As specified in `scripts/jenkins/Jenkinsfile:660`:
```
  // Build Elektra on a minimal Debian Bullseye Docker image
  // specify NODEP and see if plugins are properly included/excluded
  tasks << buildAndTest(
    "debian-bullseye-minimal-nodep",
    DOCKER_IMAGES.bullseye_minimal,
    ['PLUGINS': 'NODEP'],
    [TEST.ALL]
  )
```
The pipeline complains, that the the `dump` and `resolver` plugin aren't available:
```
CMake Error at src/plugins/CMakeLists.txt:73 (message):
  selected default storage (dump) is not selected in PLUGINS, please change
  KDB_DEFAULT_STORAGE or PLUGINS


CMake Error at src/plugins/CMakeLists.txt:81 (message):
  selected default resolver (resolver_fm_hpu_b) is not selected in PLUGINS,
  please change KDB_DEFAULT_RESOLVER or PLUGINS
```
As I left the `nodep` tag as it is, I don't know why they aren't included.
Any ideas why this might be the case @markus2330 ?

markus2330 commented:
It looks like a bug that "tested/unit" is confusing the CMake parser (not including the nodep tag). E.g.  path and profile gets added, and they both don't have a tag with /.

The fix should be fairly simply: The regex in scripts/cmake/Modules/LibAddPlugin.cmake line 286+287. need to be modified to also allow / (same as infos/provides where / is already supported).

Please have a look into what was changed when infos/provides got support for /. It would be great if the plugin inclusion of TESTED and UNIT (individual parts of tested/unit) works, too.

janldeboer committed:
scripts: Fix tag parsing

janldeboer committed:
scripts: Fix parsing tags

Janldeboer commented:
> It would be great if the plugin inclusion of TESTED and UNIT (individual parts of tested/unit) works, too.

I changed the regex accordingly, one side effect is that the `TESTED` tag my be in the category list multiple times. Is that a problem / do I have to delete potential duplications?
I can't think of a case where that leads to a problem, but maybe I overlook something.

markus2330 commented:
Something went wrong with the rebase.

Janldeboer commented:
> Something went wrong with the rebase.

Yes, I still have trouble rebasing, but now the commit history looks fine again

Janldeboer commented:
> There are many tests failing: https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4758/59/tests

Yes, as far as I can see they all came down to the same problem,  so one fix should be able to fix them all.

> Did you try to reproduce it locally? https://www.libelektra.org/tutorials/run-all-tests-with-docker

I will try that now

Janldeboer commented:
Testing it locally doesn't work for me.
When running `make -j 10` I get this:
```
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp: In member function ‘bool kconfig::FileUtility::isNextCharToken()’:
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp:44:2: error: case label value is less than minimum value for type [-Werror=switch-outside-range]
   44 |  case EOF:
      |  ^~~~
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp: In member function ‘void kconfig::FileUtility::skipLine()’:
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp:76:3: error: case label value is less than minimum value for type [-Werror=switch-outside-range]
   76 |   case EOF:
      |   ^~~~
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp: In member function ‘void kconfig::FileUtility::readUntilChar(std::ostream&, const char&)’:
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp:130:9: error: comparison is always false due to limited range of data type [-Werror=type-limits]
  130 |   if (c == EOF && isNextCharEOF ())
      |         ^
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp: In member function ‘void kconfig::FileUtility::readUntilChar(std::ostream&, const char&, const char&)’:
/home/jenkins/workspace/src/plugins/kconfig/file_utility.cpp:156:9: error: comparison is always false due to limited range of data type [-Werror=type-limits]
  156 |   if (c == EOF && isNextCharEOF ())
      |         ^
```

Janldeboer commented:
@markus2330 I've been looking into this for over an hour now but I honestly have no idea where these problems come from. Especially because the problems just came up after fixing the tag parsing.
Do you have any idea how I can continue troubleshooting?

markus2330 commented:
Looks like nodep is missing in some core plugin (probably backend plugin). Probably some bug in the parsing was introduced. You could, e.g., print what CMake parses and compare with how it should be.

kodebach commented:
I don't think `backend` is missing. The first `kdb mount` fails with the message:

```
The command /home/jenkins/workspace/libelektra_PR-4758/build directory/bin/kdb mount terminated unsuccessfully with the info:
You entered a bad name (<empty>) for a plugin!
A valid name of a plugin is either
modulename or modulename#refname
where both modulename and refname must start with a-z
and then a-z, 0-9 and underscore (_) only
Please report the issue at https://issues.libelektra.org/
```

I don't know why that happens, but if `backend` was missing the message would contain something about "Bootstrapping failed", since `backend` is already used during the bootstrap in `kdbOpen`.

An even better proof of this is that there are `testshell_markdown_` tests that do succeed:

```
testshell_markdown_blacklist .................   Passed
```

I think the issue might actually be in the C++ code of the `tools` library not in CMake. It's probably because of the `--with-recommends` of the test for `hosts`. That probably parses some additional data from the `infos/*` of and encounters a bug somewhere, which leads to it trying to load a plugin with an empty name.

kodebach commented:
Also you should only look at the `testshell_markdown_hosts` test. All the tests afterwards fail only because the test destroyed the KDB (see also #4826).

markus2330 commented:
The C++ parsing code is in src/libs/tools/src/plugin.cpp line 169-. Maybe it has problems with spaces at end or similar?

It should now be impossible that the data is out-of-date, as it directly includes plugincontract.h.

@kodebach is there actually some code you know that directly works with "global" and might have problems that it is now called "hook"? The code in src/libs/tools/src/backend.cpp GlobalPlugins::serialize seems to broken anyway?

kodebach commented:
> is there actually some code you know that directly works with "global"

I know nearly nothing about the C++ tools code. I only did the bare minimum to make `kdb mount` work in the `new-backend` PR.

> Maybe it has problems with spaces at end or similar?

AFAIK spaces at the end of lines in Markdown files would make the formatting check fail, so there shouldn't be any.

Janldeboer commented:
> That probably parses some additional data from the `infos/*` of and encounters a bug somewhere, which leads to it trying to load a plugin with an empty name.

I looked into this before, but couldn't see anything wrong so far.
What I did find suspicious, is that the `infos/status` and `infos/provides` get mixed up before being put into `split_plugin_providers`

Here they are being combined:
```
file (READ ${CMAKE_CURRENT_SOURCE_DIR}/README.md contents)
	string (REGEX MATCH "- +infos/status *= *([-a-zA-Z0-9 ]*)" CATEGORIES "${contents}")
	string (REGEX REPLACE "- +infos/status *= *([-a-zA-Z0-9 ]*)" "\\1" CATEGORIES "${CATEGORIES}")
	string (REGEX REPLACE " " ";" CATEGORIES "${CATEGORIES}")

	string (REGEX MATCH "- +infos/provides *= *([a-zA-Z0-9/ ]*)" PROVIDES "${contents}")
	string (REGEX REPLACE "- +infos/provides *= *([a-zA-Z0-9/ ]*)" "\\1" PROVIDES "${PROVIDES}")
	string (REGEX REPLACE " " ";" PROVIDES "${PROVIDES}")
	list (APPEND CATEGORIES "ALL" "${PROVIDES}")

	split_plugin_providers (CATEGORIES)
```
and than the providers are being split by this macro:
```
macro (split_plugin_providers PROVIDES)
	foreach (PROVIDER "${${PROVIDES}}")
		string (REGEX MATCH "([a-zA-Z0-9]+)/([a-zA-Z0-9]+)" PROVIDER_PARTS "${PROVIDER}")
		string (LENGTH "${PROVIDER_PARTS}" PROVIDER_PARTS_LENGTH)
		if (PROVIDER_PARTS_LENGTH GREATER 0)
			string (REGEX REPLACE "([a-zA-Z0-9]+)/([a-zA-Z0-9]+)" "\\1;\\2" PROVIDER_PARTS "${PROVIDER_PARTS}")
			list (APPEND ${PROVIDES} "${PROVIDER_PARTS}")
		endif ()
	endforeach ()
endmacro ()
```
While this seems to be a potential source of errors, I couldn't see why exactly it would be:
No category should be parsed as a provider, as all the `/` from the categories are being replaced by `;` now.

The prints that I put in there might help to check that.

kodebach commented on scripts/dev/migrate-status-to-features:
```diff
@@ -0,0 +1,88 @@
+import re
```
```suggestion
#!/usr/bin/env python

import re
```

I don't think it would fix anything, but it should enable syntax highlighting on Github ;)

kodebach commented:
The more I look at it the less I understand about this bug...

It happens in both `debian-buster-minimal` and `debian-bullseye-minimal`, which are identical jobs except for the OS version. But it does not happen in `debian-bullseye-minimal-nodep`. That leads me to conclude the issues is with one of the plugin that are **not** marked as `nodep`. But no idea, how or why...

On a random guess: Have you tried undoing the changes to  `CONTRACT.ini` that use `#include`. I didn't find any code that would read `CONTRACT.ini` directly, but maybe the `#` is what's confusing things.

janldeboer committed:
scripts: Add python hashbang to file

Co-authored-by: Klemens Böswirth <23529132+kodebach@users.noreply.github.com>

Janldeboer commented:
> The more I look at it the less I understand about this bug...

I'm glad that I'm not the only one feeling like this 🥲

> On a random guess: Have you tried undoing the changes to `CONTRACT.ini` that use `#include`? I didn't find any code that would read `CONTRACT.ini` directly, but maybe the `#` is what's confusing things.

Worth a try, I'll do so now.

Janldeboer commented:
I just noticed one other thing in the contract:
```
[infos/status]
type = vector<int enum #include "src/libs/tools/include/plugincontract.h">
```
Integers might not big enough the handle the value of the `default` tag (64000).
I'm not sure what kind of integers we got here, but long was used before so I switched back to that now.

janldeboer committed:
doc: Undo includes in CONTRACT

This is mainly for debugging

Janldeboer commented:
> On a random guess: Have you tried undoing the changes to `CONTRACT.ini` that use `#include`? I didn't find any code that would read `CONTRACT.ini` directly, but maybe the `#` is what's confusing things.

Unfortunately, the same error persists.
